Django App Plugins: support for Signal Receivers

nasthagiri opened the pull request:
This PR refactors and enhances the Django App Plugin feature with support for Signal Receivers.
This allows Plugin apps to shift its dependencies on Signal Senders from code-time to runtime.

bderusha commented:
This looks great, @nasthagiri !  Looking forward to seeing more of this proliferate throughout the platform! 👍

jmbowman commented on openedx/core/djangoapps/plugins/utils.py:
```diff
@@ -10,3 +16,32 @@ def get_module_path(app_config, plugin_config, plugin_cls):
     )
+
+
+def parse_and_import_module_attr(attr_path):
```
Django has a utility function for this: https://docs.djangoproject.com/en/1.8/ref/utils/#django.utils.module_loading.import_string

nasthagiri replied:
ah.. there you are.  I was looking in the wrong place.  Thanks!  updated the code.

jmbowman commented on setup.py:
```diff
@@ -64,13 +64,15 @@
-            "grades = lms.djangoapps.grades.apps:GradesConfig",
             "ace_common = openedx.core.djangoapps.ace_common.apps:AceCommonConfig",
+            "grades = lms.djangoapps.grades.apps:GradesConfig",
+            "plugins = openedx.core.djangoapps.plugins.apps:PluginsConfig",
```
Because signals are connected in `ready()`, will any plugins which end up after this in `INSTALLED_APPS` fail to have their signals connected because that already ran?

nasthagiri replied:
When I was thinking about this, my conclusion was that since Django has multiple phases in its Startup process, by the time `ready()` is called, all the Django apps would have already been installed and included in `INSTALLED_APPS`.  Thus, the order in which the apps are in `INSTALLED_APPS` shouldn't affect the order in which their signals get connected in `ready()`.  This is possible only because Django has this multi-phase process.

Is this what you were asking?

jmbowman replied:
I guess this works because the `INSTALLED_APPS` population happens in the settings module, not the plugin app's `ready()` method.  The tricky part is that Django apps normally register their signal receivers in `ready()`; the way the plugin system is set up, plugins that come before the plugin registration app itself will have any such direct signal connections of their own set up before declared receivers are set up, and ones that come after it will set them up afterwards.  That's a bit confusing, but not necessarily a problem.

nasthagiri replied:
I suppose I don't understand where the confusion lies.  The `ready()` method gets called only *after* all apps are installed.  By then, all senders and receivers would already have been installed.  This is why Django recommends that all signal connections happen within the `ready()` method.

jmbowman commented on openedx/core/djangoapps/schedules/apps.py:
```diff
@@ -1,12 +1,22 @@
+            ProjectType.LMS: {
+                PluginSignals.RECEIVERS: [{
+                    PluginSignals.RECEIVER_FUNC_NAME: u'update_schedules_on_course_start_changed',
+                    PluginSignals.SIGNAL_PATH: u'openedx.core.djangoapps.content.course_overviews.signals:COURSE_START_DATE_CHANGED',
```
Maybe it's just me, but needing the `ClassName.` prefixes here really hurts readability, especially because there are so many of them.  I'm a little hesitant to just dump them all in the `constants` module namespace, though.  Any other ideas on how to make this a little easier to read?

nasthagiri replied:
Would it help if each Enum-class was defined in its own Python module - and we use module encapsulation instead of Class encapsulation?

nasthagiri replied:
Another idea:
```
from openedx.core.djangoapps.plugins.constants import PluginSignals as ps, ProjectType as pt

    plugin_app = {
        ps.CONFIG: {
            pt.LMS: {
                ps.RECEIVERS: [{
                    ps.RECEIVER_FUNC_NAME: u'update_schedules_on_course_start_changed',
                    ps.SIGNAL_PATH: u'openedx.core.djangoapps.content.course_overviews.signals.COURSE_START_DATE_CHANGED',
                }]
```

Does this read better?

jmbowman commented on openedx/core/djangoapps/plugins/constants.py:
```diff
@@ -0,0 +1,60 @@
+PLUGIN_APP_SETTINGS_FUNC_NAME = u'plugin_settings'
+
+
+class ProjectType(object):
```
Weren't we going to use the `enum34` backport to make these enums?

nasthagiri replied:
I need to be able to call `getattr` with these enum values.  Does the new `Enum` class have a way to get the string value of the enum field?

jmbowman replied:
Looks like it: https://docs.python.org/3/library/enum.html#programmatic-access-to-enumeration-members-and-their-attributes

nasthagiri replied:
Ok.  Not ideal for our case though.  We'll have to remember to call `.value` whenever we pass the `Enum` as a parameter to a `dict`'s `get` or to a `getattr`.  I'm not sure converting the classes to Enums actually helps much in our case where the string values are more significant than the constant names.  I'll take a stab at this and see the impact on the code.

nasthagiri replied:
Ok.  I took a stab at changing these to be `Enum`s.  However, the usage of them no longer works as designed. 
In particular, Plugin apps need to either always use their string values or always use their `Enum` values as `dict` keys.  Either that, or the plugins code needs to check for both (string and Enum) when looking up in the `dict`.  Neither is ideal and I'd rather keep these as `Class`es given these issues.

nasthagiri commented on openedx/core/djangoapps/plugins/plugin_signals.py:
```diff
@@ -0,0 +1,64 @@
+            signal.connect(
+                receiver_func,
+                sender=_get_sender(receiver_config),
+                dispatch_uid=_get_dispatch_uuid(receiver_config, receiver_func),
```
Yeah - I couldn't think of a reason when we would *not* want this safeguard.  I also favored enforcing this for now.

If we ever want to relax this in the future, we can always introduce a new receiver preference.  We would need a better name, but something like `PluginSignals.RECEIVER_NONE_DISPATCH_UID`.

nasthagiri commented on openedx/core/djangoapps/plugins/apps.py:
```diff
@@ -0,0 +1,30 @@
+        Connect plugin receivers to their signals.
+        """
+        from openedx.core.djangoapps.plugins import constants, plugin_signals
+        from django.conf import settings
```
No, they don't need to be local imports.  And yes, they can be relative imports.
I've updated the code.

nasthagiri commented:
> would it make sense to abbreviate plugin_settings.py, plugin_urls.py, and plugin_signals.py to just settings.py, etc.? Were they named this way because they don't quite match what you normally expect to see in modules with those names, or because it throws off the plugin machinery if they use the standard names without their usual content?

I named them that way for the former reason.  Since when you peek inside, you don't see what you normally would.  I thought this naming follows the `Principle of Least Surprise`.  What do you think?

nasthagiri commented:
In this iteration, I'd like to keep the file names explicit (e.g., `plugin_settings`, etc) and let the callers be explicit from where they're importing functionality.  Let's see how this framework evolves and we can revisit if we still don't like it.  I think it may evolve better than importing all public functionality into a common `__init__.py` module - but let time tell us.

nasthagiri commented:
@jmbowman I believe I've addressed the comments that are blockers for this particular PR.  The only outstanding issue is the readability of using the Enum classes.  We can address that in a subsequent PR as it's unrelated to this refactor and enhancement.  And it gives us time to come up with a better idea.
